File: src/commands/status.ts
import { HINTS } from "../lib/hints.js";
import { buildResponse, formatResponse } from "../lib/response-builder.js";
import { resolveTaskFile } from "../lib/task-utils.js";
import { loadAndValidateTask } from "../lib/validator.js";

/**
 * hal status
 *
 * Show the current task status and all steps.
 */
export function handleStatus(opts: { file?: string }): void {
	const taskPath = resolveTaskFile(opts.file);
	const task = loadAndValidateTask(taskPath);

	const steps = (task.steps ?? []).map((s) => ({
		id: s.id,
		goal: s.goal,
		status: s.status ?? "pending",
		accept: s.accept,
	}));

	const responseData: Record<string, unknown> = {
		task_path: taskPath,
		task_id: task.id,
		goal: task.goal,
		status: task.status ?? "pending",
		steps,
	};

	if (task.accept) {
		responseData.accept = task.accept;
	}

	if (task.plan) {
		responseData.plan = task.plan;
	}

	let commandsHint: string;
	const taskStatus = task.status ?? "pending";

	if (taskStatus === "done") {
		commandsHint = HINTS.STATUS_DONE;
	} else if (taskStatus === "pending") {
		if (steps.length === 0) {
			commandsHint = HINTS.STATUS_NO_STEPS;
		} else {
			commandsHint = HINTS.STATUS_PENDING;
		}
	} else {
		const currentStep = steps.find((s) => s.status === "in_progress");
		const nextPending = steps.find((s) => s.status === "pending");

		if (currentStep) {
			commandsHint = HINTS.STEP_IN_PROGRESS(currentStep.id);
		} else if (nextPending) {
			commandsHint = HINTS.STEP_DONE_NEXT(nextPending.id);
		} else {
			commandsHint = HINTS.STATUS_ADD_OR_CHECK;
		}
	}

	const response = buildResponse({
		status: "ok",
		message: `Task status: ${taskStatus}`,
		data: responseData,
		commands_hint: commandsHint,
	});

	console.log(formatResponse(response));
}

File: src/commands/status.rs
use anyhow::Result;
use std::collections::BTreeMap;

use crate::core::hints;
use crate::core::response::HalResponse;
use crate::core::task_utils::resolve_task_file;
use crate::core::validator::load_and_validate_task;
use crate::types::Status;

pub fn handle_status(file: Option<&str>) -> Result<()> {
    let task_path = resolve_task_file(file)?;
    let task = load_and_validate_task(&task_path)?;

    let task_status = task.status.as_ref().unwrap_or(&Status::Pending);
    let steps = task.steps.as_deref().unwrap_or(&[]);

    let mut data = BTreeMap::new();
    data.insert("task_path".into(), serde_yaml::Value::String(task_path.display().to_string()));
    data.insert("task_id".into(), serde_yaml::Value::String(task.id.clone()));
    data.insert("goal".into(), serde_yaml::Value::String(task.goal.clone()));
    data.insert("status".into(), serde_yaml::Value::String(task_status.to_string()));

    let steps_yaml: Vec<serde_yaml::Value> = steps.iter().map(|s| {
        let mut m = serde_yaml::Mapping::new();
        m.insert(serde_yaml::Value::String("id".into()), serde_yaml::Value::String(s.id.clone()));
        m.insert(serde_yaml::Value::String("goal".into()), serde_yaml::Value::String(s.goal.clone()));
        m.insert(serde_yaml::Value::String("status".into()), serde_yaml::Value::String(s.status.as_ref().unwrap_or(&Status::Pending).to_string()));
        serde_yaml::Value::Mapping(m)
    }).collect();
    data.insert("steps".into(), serde_yaml::Value::Sequence(steps_yaml));

    let hint = match task_status {
        Status::Done => hints::STATUS_DONE.to_string(),
        Status::Pending if steps.is_empty() => hints::STATUS_NO_STEPS.to_string(),
        Status::Pending => hints::STATUS_PENDING.to_string(),
        _ => {
            if let Some(s) = steps.iter().find(|s| s.status.as_ref() == Some(&Status::InProgress)) {
                hints::step_in_progress(&s.id)
            } else if let Some(s) = steps.iter().find(|s| s.status.as_ref().unwrap_or(&Status::Pending) == &Status::Pending) {
                hints::step_done_next(&s.id)
            } else {
                hints::STATUS_ADD_OR_CHECK.to_string()
            }
        }
    };

    let response = HalResponse::new("ok", &format!("Task status: {}", task_status))
        .with_data(data)
        .with_hint(&hint);
    println!("{}", response.to_yaml());
    Ok(())
}
